File: MapDesign/MapDesign.cs
using System.Text;

namespace MapDesignLibrary;

/// <summary>
/// Represents a valid dungeon design, main output objects are the TileMatrix, DisplayMatrix and 
/// ImageMap fields.
/// </summary>
public class MapDesign
{
    // Represents each section of the map and its current connections *3D*
    internal List<List<SortedSet<string>>> ConnectionMatrix { get; } = [];
    // Represents sections in the map that should be protected from random seeding
    internal SortedSet<(int row, int col)> SafetyBuffer = [];
    // Represent sections chosen to have a random tile placed in it
    internal SortedSet<(int row, int col)> SeededSections = [];
    internal int RequiredSeeds = 0;

    // Represents Entry point & stairs into the dungeon, -1 indicates not used/assigned
    internal (int row, int col) Entrance = (-1, -1);
    internal (int row, int col) StairsUp = (-1, -1);
    internal (int row, int col) StairsDown = (-1, -1);

    /* ! Edges and corners require special handling ! */
    internal Dictionary<string, List<(int row, int col)>> Edges = [];

    /* ! Main Outputs ! */
    // Represents each tile id that can fill the connection matrix above *3D*
    public List<List<List<int>>> TileMatrix { get; } = [];
    // Represents the tiles chosen for display *2D*
    public List<List<int>> DisplayMatrix { get; } = [];
    // Holds the associated filename for the output matrixes
    public Dictionary<int, string> ImageMap { get; } = [];

    /// <summary>
    /// Returns the map design as string, providing all the details needed to verify
    /// designs.
    /// </summary>
    /// <returns>String representation of the map</returns>
    public override string ToString()
    {
        StringBuilder builder = new();

        builder.AppendLine();
        builder.AppendLine("Connection Matrix");
        builder.AppendLine("---------------------------------------------------");
        builder.AppendLine();

        foreach (var row in ConnectionMatrix)
        {
            builder.Append('[');
            foreach (var col in row)
            {
                builder.Append('[');
                builder.Append(string.Join(", ", col));
                builder.Append(']');
            }
            builder.AppendLine("]\n");
        }

        builder.AppendLine();
        builder.AppendLine("Details: Seeds, Buffer, Entry|Exit");
        builder.AppendLine("---------------------------------------------------");
        builder.AppendLine();

        builder.AppendLine($"Entrance = {Entrance.row}, {Entrance.col} : (row, col)");
        builder.AppendLine($"Stairs Up = {StairsUp.row}, {StairsUp.col} : (row, col)");
        builder.AppendLine($"Stairs Down = {StairsDown.row}, {StairsDown.col} : (row, col)");

        builder.Append("Safety Buffer (row, col) = ");
        foreach (var (row, col) in SafetyBuffer)
        {
            builder.Append($"({row}, {col})");
        }
        builder.AppendLine();

        builder.AppendLine($"Required Seeds = {RequiredSeeds}");

        builder.Append("Seeded Sections (row, col) = ");
        foreach (var (row, col) in SeededSections)
        {
            builder.Append($"({row}, {col})");
        }

        builder.AppendLine();
        builder.AppendLine("Tile Matrix");
        builder.AppendLine("---------------------------------------------------");
        builder.AppendLine();

        foreach (var row in TileMatrix)
        {
            builder.Append('[');
            foreach (var col in row)
            {
                builder.Append('[');
                builder.Append(string.Join(", ", col));
                builder.Append(']');
            }
            builder.AppendLine("]\n");
        }

        builder.AppendLine();
        builder.AppendLine("Display Matrix");
        builder.AppendLine("---------------------------------------------------");
        builder.AppendLine();

        foreach (var row in DisplayMatrix)
        {
            builder.Append('[');
            builder.Append(string.Join(", ", row));
            builder.AppendLine("]\n");
        }

        return builder.ToString();
    }
}

File: dungeon/src/designs/MapDesign.java
package designs;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;

import models.MapCoordinate;

/**
 * This class represents a dungeon map design made to create a map image for
 * users. Main outputs used are the display matrix, valid id matrix, and image
 * map.
 */
public class MapDesign {

    ArrayList<ArrayList<TreeSet<String>>> connectionMatrix = new ArrayList<>();
    TreeSet<MapCoordinate> seededSections = new TreeSet<>();
    TreeSet<MapCoordinate> safetyBuffer = new TreeSet<>();
    HashMap<String, ArrayList<MapCoordinate>> edgeMap = new HashMap<>();
    MapCoordinate entrance;
    MapCoordinate stairsDown;
    int requiredSeeds;

    // Main outputs
    ArrayList<ArrayList<Integer>> displayMatrix = new ArrayList<>();
    ArrayList<ArrayList<HashSet<Integer>>> validIdMatrix = new ArrayList<>();
    HashMap<Integer, String> imageMap = new HashMap<>();

    /**
     * Returns the display matrix of the map design.
     * 
     * @return a 2d ArrayList of the tile ids choosen to be initialy displayed
     */
    public ArrayList<ArrayList<Integer>> getDisplayMatrix() {
        return displayMatrix;
    }

    /**
     * Returns the 3d array/set container which holds the set of ids for each
     * coordinate in the map design.
     * 
     * @return 3d ArrayList of ArrayList of HashSet of ids (Integer)
     */
    public ArrayList<ArrayList<HashSet<Integer>>> getValidIdMatrix() {
        return validIdMatrix;
    }

    /**
     * Returns the map of ids to images for the map, used for display to user and
     * image merging.
     * 
     * @return HashMap of ids(key) and image filenames(value)
     */
    public HashMap<Integer, String> getImageMap() {
        return imageMap;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();

        builder.append("----Connection Matrix----\n");
        for (ArrayList<TreeSet<String>> row : connectionMatrix) {
            builder.append("[");
            for (TreeSet<String> col : row) {
                builder.append("[");
                for (String connection : col) {
                    builder.append(connection);
                    builder.append(" ");
                }
                builder.append("]");
            }
            builder.append("]\n");
        }

        builder.append("\n\n----Descriptions----\n");
        builder.append("Seeded Sections: (row,col)");
        for (MapCoordinate coordinate : seededSections) {
            builder.append(coordinate.toString());
            builder.append(" ");
        }

        builder.append("\nSaftey Buffer: (row,col)");
        for (MapCoordinate coordinate : safetyBuffer) {
            builder.append(coordinate);
            builder.append(" ");
        }

        if (entrance != null) {
            builder.append("\nEntrance: (row,col)");
            builder.append(entrance.toString());
        }

        if (stairsDown != null) {
            builder.append("\nStairs Up: (row,col)");
            builder.append(stairsDown.toString());
        }

        builder.append("\n\n----ID Display Matrix----\n");
        for (ArrayList<Integer> row : displayMatrix) {
            builder.append("[");
            for (Integer colInt : row) {
                builder.append(colInt);
                builder.append(" ");
            }
            builder.append("]\n");
        }

        builder.append("\n\n----Valid ID Matrix----\n");
        for (ArrayList<HashSet<Integer>> row : validIdMatrix) {
            builder.append("[");
            for (HashSet<Integer> col : row) {
                builder.append("[");
                for (Integer id : col) {
                    builder.append(id);
                    builder.append(" ");
                }
                builder.append("]");
            }
            builder.append("]\n");
        }

        return builder.toString();
    }
}
